### PR TITLE
Menu button updates for table

### PR DESCRIPTION
**Pull request by mollykreis**
# Pull Request

## 🤨 Rationale

Some changes need to be made to the menu button before it can be used within the table for the action menu. Specifically, those changes are:
- adding a `beforetoggle` event
- having the menu button find the slotted menu when it is nested within multiple slots

See #956 for more details.

## 👩‍💻 Implementation

- Add `beforetoggle` event to the menu button that is fired before the menu opens or closes
- Add Angular & Blazor support for the `beforetoggle` event
- Change the logic for finding the slotted menu. We can no longer use a filter in the template to find slotted elements with a role of `menu` because this won't match `slot` elements. Instead, the template assigns all slotted elements to the `slottedMenus` array, and the getter for `menu` finds the menu, which may be nested within `slot` elements.

## 🧪 Testing

- Manually tested that the `beforetoggle` event handler works correctly in Blazor & Angular
- Added auto tests for the `beforetoggle` event works correctly for the web components
- Basic manual testing in Storybook that the component still behaves as expected
- Added tests that the focus & menu opening behavior works correctly when the menu is nested in an additional slot

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

**Review by jattasNI — APPROVED** [review]

File: packages/nimble-components/src/menu-button/specs/README.md [hunk 1]
```diff
@@ -84,6 +84,9 @@ _Methods_
 
 _Events_
 
+-   `beforetoggle` (event) - event fired before the opened state has changed. The event detail contains:
+    -   `newState` - boolean - The value of `open` on the menu button that the element is transitioning in to.
+    -   `oldState` - boolean - The value of `open` on the menu button that the element is transitioning out of.
 -   `open-change` (event) - event for when the opened state has changed
```
Comment (rajsite) [inline packages/nimble-components/src/menu-button/specs/README.md:90]: From [spec comments](https://github.com/ni/nimble/pull/956#discussion_r1073042699) I think we aligned on having `beforetoggle` and `toggle` events instead of `open-change`. Do we plan to do that in the future? Should we remove open-change now or only when switching to toggle?

File: packages/nimble-components/src/menu-button/index.ts [hunk 2]
```diff
@@ -162,16 +162,52 @@ export class MenuButton extends FoundationElement implements ButtonPattern {
...
+
+        let currentItem = this.slottedMenus[0];
+        while (currentItem) {
+            if (currentItem.getAttribute('role') === 'menu') {
+                return currentItem;
+            }
+
+            if (currentItem?.nodeName === 'SLOT') {
+                currentItem = (
+                    currentItem as HTMLSlotElement
```
Comment (rajsite) [inline packages/nimble-components/src/menu-button/index.ts:203]: Can make a simple private type guard with a type predicate to clean-up this check: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

File: packages/nimble-components/src/menu-button/index.ts [hunk 3]
```diff
@@ -162,16 +162,52 @@ export class MenuButton extends FoundationElement implements ButtonPattern {
...
+                ).assignedNodes()[0] as HTMLElement;
```
Comment (rajsite) [inline packages/nimble-components/src/menu-button/index.ts:204]: if we can't avoid the `as` cast then we should try and represent it better, maybe `as HTMLElement | undefined`?

I quickly scanned the tests and don't think I saw them but do we have tests for edge cases like the most nested slot doesn't have children, has child text nodes, or has child non-menu elements? (are those useful tests?)

File: packages/nimble-components/src/menu-button/specs/README.md (hunk 1 above)
Comment (mollykreis) [reply packages/nimble-components/src/menu-button/specs/README.md:90]: I was hesitant to make that change since it would be a breaking change. I was pretty sure we were using that event in SLE, but it turns out we aren't. I'll go ahead and rename it as part of this PR.

**Review by atmgrifter00 — APPROVED** [review]

File: packages/nimble-components/src/menu-button/index.ts (hunk 3 above)
Comment (mollykreis) [reply packages/nimble-components/src/menu-button/index.ts]: I changed this code slightly to use `instanceof`, which allowed me to get rid of the `as` cast.

I also updated the tests to include a cases where a menu wasn't provided (both in the nested slot case and the standard case). I'm only running some tests in those cases because a lot of the focus behavior doesn't apply to not having a menu. I'm not sure how useful the tests are because the menu button isn't even functional in those cases since it doesn't have a menu, but it at least ensures we don't throw an exception if the timing works out that the menu doesn't exist yet when the button first opens.

File: angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts [hunk 4]
```diff
@@ -46,14 +47,5 @@ export class NimbleMenuButtonDirective {
         this.renderer.setProperty(this.elementRef.nativeElement, 'open', toBooleanProperty(value));
     }
 
-    @Output() public openChange = new EventEmitter<boolean>();
```
Comment (mollykreis) [inline angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: Note: I removed this `Output` and didn't replace it with a `toggle` `Output` because it causes problems when the Angular `Output` and the native event have the exact same name. The specific issue is that a client app gets their handler called twice -- once triggered by the native element and once triggered by the directive's `Output`. I thought it would be best to simply not have the Angular directive do anything with the event (other than export the type of the event detail).

**Review by msmithNI — APPROVED** [review]

File: angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts (hunk 4 above)
Comment (rajsite) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: Does the Angular `@Output` gives better typing / inspection for Angular users than them binding to the native events? Maybe if we expect Angular clients to use the `beforetoggle` and `toggle` events then we could stopImmediatePropagation of the native event in the `@HostListener` binding so client bindings with the same name will only be hit once by `@Output` emitted version?

File: packages/nimble-components/src/menu-button/index.ts [hunk 5]
```diff
@@ -162,16 +170,61 @@ export class MenuButton extends FoundationElement implements ButtonPattern {
...
+        const eventDetail: MenuButtonToggleEventDetail = {
+            oldState: this.open,
+            newState: newValue
+        };
+        this.$emit('beforetoggle', eventDetail);
+
+        this.open = newValue;
+    }
+
     private get menu(): HTMLElement | undefined {
```
Comment (rajsite) [inline packages/nimble-components/src/menu-button/index.ts:195]: nit / pref: this is doing a fair amount of work at this point so I'd probably say lets go ahead and make this a `getMenu()` method instead

**Review by rajsite — APPROVED** [review]

File: angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts (hunk 4 above)
Comment (mollykreis) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: The difference in Angular is that `$event` is of type `Event` if nothing is done in the directive and is of type `MenuButtonToggleEventDetail` if we pull the details out in the directive.

If we do nothing in the directive, an Angular user would do something like this in their event handler:
```ts
public onMenuButtonToggle(event: Event) {
    const toggleDetail= (event as CustomEvent).detail as MenuButtonToggleEventDetail;
    // do something with toggleDetail.newState and toggleDetail.oldState
}
```

The typing for an Angular client is better if we do the `as MenuButtonToggleEventDetail` cast in the directive. It just means that they lose access to any other configuration on the original `Event`, but I think that's fine. I'll expose `toggle` and `beforeToggle` as `@Output`s in the directive. Only `toggle` will need to stop propagation since the names are the same for that event/output unlike `before-toggle` and `beforeToggle`. I lean towards not stopping propagation in the `before-toggle` event handler because we don't do that for any of our other handlers within nimble.

Comment (rajsite) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: > I lean towards not stopping propagation in the before-toggle event handler because we don't do that for any of our other handlers within nimble.

Yea I guess that's because none of the other ones have the name conflict so we haven't run into it. It does seem a bit odd that a client could bind to either `beforetoggle` or `beforeToggle` but we can let it run for a while and see if we want to prevent those from leaking through.


> t just means that they lose access to any other configuration on the original Event, but I think that's fine

Can we make an interface that represents the Event to export from components? Then we can keep the normal event data in there. Probably don't need any of the `as` casts either

```ts
interface MenuButtonToggleEvent extends CustomEvent {
   details: MenuButtonToggleEventDetail;
}
```

Comment (rajsite) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: > My personal preference is to leave the event listeners off the directive.

If we don't have an immediate need for them we can kick the can down the road and not expose them yet. If we do need them I think it's worth thinking about (I'd be concerned if clients listen to the event directly which we don't have a great way to type enforce compared to defining `@Output`), but it could be in a follow-up PR to not block action menu work.

Comment (mollykreis) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: > Can we make an interface that represents the Event to export from components? Then we can keep the normal event data in there. Probably don't need any of the as casts either

We can make that interface, but it doesn't completely avoid the need for `as` casts because `$event` will be `Event` not `MenuButtonToggleEvent`
```ts
public onMenuButtonToggle(event: Event) {
    const customEvent = event as MenuButtonToggleEvent;
    // do something with customEvent.detail.newState and customEvent.detail.oldState
}
```

>If we don't have an immediate need for them we can kick the can down the road and not expose them yet.

I'm not aware of a need for them. I do think deferring this work is the best option because I think we need to take a holistic approach to deciding which events get turned into Angular Outputs and which we provide event types for. Right now I don't think we're consistent in what we expose in Angular, and I think creating an interface would only increase the inconsistency.

Comment (rajsite) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: Sounds good to me! Maybe a tech debt issue for capturing some of the concerns we want to clean-up would help track it.

Comment (mollykreis) [reply angular-workspace/projects/ni/nimble-angular/src/directives/menu-button/nimble-menu-button.directive.ts:49]: Created #999
